src/binary/load_kernel.rs

```diff
@@ -496,6 +504,44 @@ where
 
         Ok(())
     }
+
+    /// Mark a region of memory indicated by a GNU_RELRO segment as read-only.
+    ///
+    /// This is a security mitigation used to protect memory regions that
+    /// need to be writable while applying relocations, but should never be
+    /// written to after relocations have been applied.
+    fn handle_relro_segment(&mut self, program_header: ProgramHeader) {
+        let start = self.virtual_address_offset + program_header.virtual_addr();
+        let end = start + program_header.mem_size();
+        let start = VirtAddr::new(start);
+        let end = VirtAddr::new(end);
+        let start_page = Page::containing_address(start);
+        let end_page = Page::containing_address(end - 1u64);
```

Comment:
It is possible that the segment shares a page with other segments that should stay writable? In that case, marking the whole page as read-only would lead to issues...

Reply:
I don't think so, but I could find any spec that says so.
The spec only says:
> The array element specifies the location and size of a segment which may be made read-only after relocation shave been processed.

(https://refspecs.linuxbase.org/LSB_3.1.1/LSB-Core-generic/LSB-Core-generic.html)

I found different implementations:
glibc's runtime linker seems to round down the start as well as the end (https://sourceware.org/git/?p=glibc.git;a=blob;f=elf/dl-reloc.c;h=756bf950f6f98aa04e4ee23d202f27c26775584e;hb=refs/heads/master#l353) whereas android's runtime linker seems to round the start down and the end up (https://android.googlesource.com/platform/bionic/+/master/linker/linker_phdr.cpp#888). 

LLVM's lld always makes sure that the end of the segment is always page aligned (https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/lld/ELF/Writer.cpp#L2633-L2638), but doesn't say anything about the start.

I think the current implementation is fine.

Reply:
Ok, thanks for checking!